### PR TITLE
boot: support device change

### PR DESCRIPTION
**Pull request by bboozzoo**
Add helper for switching device model with the intention to use use in devicestate during remodel.

### CODE REVIEW

**Review by pedronis — CHANGES_REQUESTED** [review]
a question, I think we also need tests where old and new are very close, not completely different

File: boot/modeenv.go [hunk 1]
```diff
@@ -376,6 +376,27 @@ func (m *Modeenv) TryModelForSealing() secboot.ModelForSealing {
 	}
 }
 
+func (m *Modeenv) useThisModel(model *asserts.Model) {
```
[the remark this answers is not in the document]
Comment (bboozzoo) [reply boot/modeenv.go:379]: i went with setModel for consistency with setTryModel

File: boot/model.go [hunk 2]
```diff
@@ -46,16 +46,14 @@ func DeviceChange(from Device, to Device) error {
...
 	modified := false
-	if m.TryModel != newModel.Model() {
+	if modelUniqueID(m.TryModelForSealing()) != modelUniqueID(newModel) {
 		// we either haven't been here yet, or a reboot occurred after
 		// try model was cleared and modeenv was rewritten
 		m.setTryModel(newModel)
 		modified = true
 	}
-	if m.Model != oldModel.Model() {
+	if modelUniqueID(m.ModelForSealing()) != modelUniqueID(oldModel) {
```
Comment (pedronis) [inline boot/model.go:56]: are there new unit tests for this change?

Comment (bboozzoo) [reply boot/model.go:56]: I've updated one of the new unit tests to cover this code path

File: boot/model.go [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+		// modeenv has not been written and still contains both the old
+		// and a new model, but the model file has been modified,
+		// restore the original model file
+		if restoreErr := writeModelToUbuntuBoot(from.Model()); restoreErr != nil {
+			return fmt.Errorf("%v (restoring model failed: %v)", err, restoreErr)
+		}
+		// however writing modeenv failed, so trying to clear the model
+		// and write it again could be pointless, let the failure
+		// percolate up the stack
+		return err
```
Comment (pedronis) [inline boot/model.go:109]: we are hitting the double failure path and not this one

File: boot/model_test.go [hunk 4]
```diff
@@ -0,0 +1,1069 @@
...
+	// system is encrypted
+	s.stampSealedKeys(c, s.rootdir)
+
+	// mangle the modeenv to observe the right model content restored for
+	// resealing
+	m, err := boot.ReadModeenv("")
+	c.Assert(err, IsNil)
+	m.Model = "foo"
+	m.TryModel = "bar"
+	m.TryBrandID = "baz"
```
Comment (pedronis) [inline boot/model_test.go:503]: this kind of test tweak is a bit strange to me, the tests are mostly trying to be realistic but this bit is not oviously so.

My original thinking was to have a test where only the ModelSignKeyID changes and we get back to this device change after rebooting after the modeenv Write at line 99...

**Review by pedronis — COMMENTED** [review]
lgtm, still some comments about the tests

File: boot/model_test.go [hunk 5]
```diff
@@ -1067,3 +1083,158 @@ func (s *modelSuite) TestDeviceChangeUnhappyFailReseaWithSwappedModelMockedWrite
...
+		resealKeysCalls++
+		c.Logf("reseal key call: %v", resealKeysCalls)
+		m, err := boot.ReadModeenv("")
+		c.Assert(err, IsNil)
+		currForSealing := boot.ModelUniqueID(m.ModelForSealing())
+		tryForSealing := boot.ModelUniqueID(m.TryModelForSealing())
+		// timeline & calls:
+		// 1, 2 - pre reboot, run & recovery keys, try model set
+		// 3 - run key, after model file has been modified, try model cleared
+		// 4 - recovery key, model file has been modified, triggers reboot, try model cleared
```
Comment (pedronis) [inline boot/model_test.go:1132]: I didn't think of it before,  but triggers reboot make it sound like it's expected, instead we are simulating unexpected reboots, no? so maybe everywhere

triggers reboot/triggers 'reboot'  ->  unexpected reboot is triggered

**Review by pedronis — APPROVED** [review]
thank you for the test changes,  a remark about the comments that I didn't consider before

File: boot/model_test.go [hunk 6]
```diff
@@ -0,0 +1,1243 @@
...
+
+	newUC20dev := boottest.MockUC20Device("", boottest.MakeMockUC20Model(map[string]interface{}{
+		"model":             "my-model-uc20",
+		"brand-id":          "my-brand",
+		"grade":             "dangerous",
+		"sign-key-sha3-384": newKeyID,
+	}))
+
+	restore := boot.MockSeedReadSystemEssential(func(seedDir, label string, essentialTypes []snap.Type, tm timings.Measurer) (*asserts.Model, []*seed.Snap, error) {
+		s.readSystemEssentialCalls++
```
Comment (mvo5) [inline boot/model_test.go:1111]: (nitpick^2) I was initially a bit puzzled as this looks super similar to the original mocking but of course it uses a different model here. I suppose we could reuse the mocking by simply `s.oldUc20dev = boottest.MockU...` and `s.newUc20dev = ...`.

File: boot/model_test.go [hunk 7]
```diff
@@ -0,0 +1,1243 @@
...
+func (s *modelSuite) TestDeviceChangeUnhappyMockedWriteModelToBootBeforeModelSwap(c *C) {
+	s.testDeviceChangeUnhappyMockedWriteModelToBoot(c, unhappyMockedWriteModelToBootTestCase{
+		expectedErr: "cannot write new model file: mocked fail in write model to boot",
+	})
+}
+
+func (s *modelSuite) TestDeviceChangeUnhappyMockedWriteModelToBootAfterModelSwapFailingRestore(c *C) {
+	// writing modeenv after placing new model file on disk fails, and so
+	// does restoring of the old model
+	if os.Getuid() == 0 {
```
Comment (mvo5) [inline boot/model_test.go:979]: Thanks for this

**Review by mvo5 — APPROVED** [review]
Thank you, some nitpicks but looks great (and like it was really hard to write). Thanks for the detailed tests as well.

**Pull request merged**